##### Experimental script Aug27, 2020 #####

##### Create Seurat object - Standard pipeline #####
library(Seurat)
library(cowplot)

AASKN_3w.data <- Read10X(data.dir = "E:/AA scTimeCourse/Shallow_seq/Analysis/GEX/SKN/AASKN_3w/")
AASKN_old.data <- Read10X(data.dir = "E:/AA scTimeCourse/Shallow_seq/Analysis/GEX/SKN/AASKN_old/")
UGSKN_3w.data <- Read10X(data.dir = "E:/AA scTimeCourse/Shallow_seq/Analysis/GEX/SKN/UGSKN_3w/")

AASKN_3w <- CreateSeuratObject(counts = AASKN_3w.data, project = "AA3w")
AASKN_old <- CreateSeuratObject(counts = AASKN_old.data, project = "AAold")
UGSKN_3w <- CreateSeuratObject(counts = UGSKN_3w.data, project = "UG3w")

AASKN_3w[["percent.mt"]] <- PercentageFeatureSet(AASKN_3w, pattern = "^mt-")
AASKN_old[["percent.mt"]] <- PercentageFeatureSet(AASKN_old, pattern = "^mt-")
UGSKN_3w[["percent.mt"]] <- PercentageFeatureSet(UGSKN_3w, pattern = "^mt-")

VlnPlot(AASKN_3w, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(AASKN_old, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(UGSKN_3w, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

AASKN_3w <- subset(AASKN_3w, subset = nFeature_RNA > 200 & nFeature_RNA < 2000 & percent.mt < 5)
AASKN_old <- subset(AASKN_old, subset = nFeature_RNA > 200 & nFeature_RNA < 2000 & percent.mt < 5)
UGSKN_3w <- subset(UGSKN_3w, subset = nFeature_RNA > 200 & nFeature_RNA < 2000 & percent.mt < 5)

#AASKN_3w_norm <- NormalizeData(AASKN_3w)
#AASKN_old_norm <- NormalizeData(AASKN_old)
#UGSKN_3w_norm <- NormalizeData(UGSKN_3w)

#AASKN_old_downsample <- AASKN_old[, sample(colnames(AASKN_old, size = 1500, replace=F)]

##### NOTE #####
#Barcode has to be the same as in TCR filtered contig file, with the same prefix
#add.cell.ids is important. The final format has to be XXX_XXX_Barcode (Sample_ID_Barcode)

all.merge <- merge(AASKN_3w, y=c(AASKN_old, UGSKN_3w), add.cell.ids = c("AA3w_SKN", "AAold_SKN", "UG3w_SKN"), project = "MERGE", merge.data = TRUE)

all.list <- SplitObject(all.merge, split.by = "orig.ident")

all.list <- lapply(X = all.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
  })

all.anchors <- FindIntegrationAnchors(object.list = all.list, dims = 1:25)

all.combined <- IntegrateData(anchorset = all.anchors, dims = 1:25)

DefaultAssay(all.combined) <- "integrated"
all.combined <- ScaleData(all.combined, verbose = FALSE)
all.combined <- RunPCA(all.combined, npcs = 30, verbose = FALSE)

all.combined_umap <- RunUMAP(all.combined, reduction = "pca", dims = 1:25)
all.combined_umap <- FindNeighbors(all.combined_umap, reduction = "pca", dims = 1:25)
all.combined_umap <- FindClusters(all.combined_umap, resolution = 0.7)

DimPlot(all.combined_umap, reduction = "umap", split.by = "orig.ident")


##### scRepertoire #####

library(scRepertoire)
library(ggplot2)

#Load filtered contig csv
contig_dir <- "E:/AA scTimeCourse/Shallow_seq/Analysis/TCR/SKN/"
setwd(contig_dir)
tissue = "SKN"

#####TCR Barcode removal function#####
#No longer needed - Aug 28, 2020
tcr_barcoder <- function(df, prefix,  trim="\\-1"){
  
  df$barcode <- gsub(trim, "", df$barcode)
  df$barcode <- paste0(prefix, df$barcode)
  
  df
}

##### Create list of csv of for the filtered contig files #####

csv <- list()

listcsv <- list.files(pattern = "*.csv")

for (i in 1:length(listcsv)) {
  csv[[i]] <- read.csv(listcsv[i], stringsAsFactors = FALSE)
}

#Change cells = "T-GD" will give TCRG, TCRD.
#samples = first part of prefix, ID =  second part of prefix.
#Final row name format is sample_ID_Barcode, must match seurat object
combined <- combineTCR(csv, samples = names(csv), ID = rep(tissue, length(csv)), cells ="T-AB")

##### Clonotype Visualization #####

#Quantify contigs with histogram and export a table
#gene+nt is the best definition for clonotype
#clonotype by uniqte "gene" is the most sensitive, by AA and nt are moderate (b/c each codon can be generated by multiple genes)
#Scale = T, percentage of unique clonotypes in the total repertoire of each sample.
#Scale = F, number of unique clonotypes in each sample.
quantContig(combined, cloneCall="gene+nt", scale = TRUE) #Plot
quantContig_output <- quantContig(combined, cloneCall="gene+nt", scale = TRUE, exportTable = TRUE) #Export table

#Quantify by ID and call genes instead
quantContig(combined, cloneCall="gene", group = "ID", scale = TRUE) 
quantContig(combined, cloneCall="gene", scale = TRUE)

#Plot the number or clonotypes with different level of abundance
abundanceContig(combined, cloneCall = "gene", scale = F, exportTable = TRUE)

#Visualize the amino acid or nucleotide length of the CDR3
#chains "combined" or "single"
#"Combined" will return length distribution of NA with TRA/TRB, or multiple TRA/B for one cell
#"Single" will only return TRA and TRB separately
lengthContig(combined, cloneCall="aa", chains = "combined")
lengthContig(combined, cloneCall="aa", chains = "single")

#Compare clonotype between samples or tissue types
#Return the top 5 clonotypes with highest frequency
compareClonotypes(combined, numbers = 5, samples = names(combined), cloneCall="aa", graph = "alluvial")

#Explore clonotype homeostasis, add exportTable = TRUE to output table
plot1 <- clonalHomeostasis(combined, cloneCall = "gene")
plot2 <- clonalHomeostasis(combined, cloneCall = "aa")
x <- gridExtra::grid.arrange(plot1, plot2, ncol=1)

##### Seurat Integration #####

seurat <- combineExpression(combined, all.combined_umap, cloneCall="gene", groupBy = "sample")

colorblind_vector <- colorRampPalette(c("#FF4B20", "#FFB433", "#C6FDEC", "#7AC5FF", "#0348A6"))

# Clonotype expansion visualization - setup
slot(seurat, "meta.data")$cloneType <- factor(slot(seurat, "meta.data")$cloneType, 
                                              levels = c("Hyperexpanded (100 < X <= 500)", "Large (20 < X <= 100)", 
                                                         "Medium (5 < X <= 20)", "Small (1 < X <= 5)", 
                                                         "Single (0 < X <= 1)", NA))

# Hover plot setup

plot <- DimPlot(seurat, group.by = "cloneType") +
  scale_color_manual(values = colorblind_vector(5), na.value="grey")

HoverLocator(plot = plot, information = FetchData(seurat, vars = c("orig.ident", "CTaa", "barcode")))

# Other examples
seurat <- highlightClonotypes(seurat, cloneCall= "aa", sequence = c("CAAEGNYNQGKLIF_CASSGQGANTLYF", "NA_CASSPGQYEQYF"))
plot <- DimPlot(seurat, group.by = "highlight")




